### PR TITLE
Decouple the gbz80 instruction implementation

### PR DESCRIPTION
**Pull request by Rsysz**
However, without `gameboy.c`, i can't make the file.

### CODE REVIEW

**Comment by jserv** [comment]
One of the 7 rules of a [great commit message](https://chris.beams.io/posts/git-commit/):
* Limit the subject line to 50 characters

Change the subject of git commit messages.

File: cpu.h [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#include "gameboy.h"
+
+uint8_t __gb_execute_cb(struct gb_s *gb);
+
+/* Internal function used to step the CPU */
+void __gb_step_cpu(struct gb_s *gb);
```
Comment (jserv) [inline cpu.h]: Both and `__gb_execute_cb ` and `____gb_step_cpu` are internal functions, and you should keep them "static."
The `cpu.h` should provide the public function `gb_frame`, which calls the above static functions.

**Review by jserv — CHANGES_REQUESTED** [review]
The emulator context `struct gb_s` should be moved out to separate header. Let's say `core.h`, which contains the necessary definitions.

File: core.h [hunk 2]
```diff
@@ -0,0 +1,284 @@
+#include <stdint.h>
```
Comment (jserv) [inline core.h:4]: Add the following code in the beginning of this header:
```cpp
#ifndef GAMEBOYEMU_CORE
#define GAMEBOYEMU_CORE
```

In the end of this header, of course, add the corresponding line:
```cpp
#endif
```

**Comment by jserv** [comment]
> However, without `gameboy.c`, i can't make the file.

You have to modify `Makefile` in order to build the proposed changes.

**Review by jserv — CHANGES_REQUESTED** [review]
Function "gb_run_frame" should be moved to file "cpu.c"

**Comment by jserv** [comment]
Use `git rebase -i` to rework the existing git commits. I expect the following commits:
1. Move emulator context to dedicated header
2. Move cpu-specific implement out of single header

That is, you should merge the following commits into change "Move cpu-specific implement out of single header":
* Move __gb_step_cpu to dedicated cpu header
* Move __gb_execute_cb to dedicated cpu header
* Move gb_run_frame to dedicated cpu source

In other words, so-called "functional change" refers to the changes with relevant sequences.

Reference: https://backlog.com/git-tutorial/tw/stepup/stepup7_6.html

**Comment by Rsysz** [comment]
Should I move `__gb_read`, `__gb_write` and `__gb_draw_line` out of `gameboy.h` ?
if so, can i move `__gb_read`, `__gb_write` to `mmu.h`, `mmu.c`  & remove static mark, 
                 move `__gb_draw_line` to `cpu.h` ?

**Comment by jserv** [comment]
> Should I move `__gb_read`, `__gb_write` and `__gb_draw_line` out of `gameboy.h` ?
> if so, can i move `__gb_read`, `__gb_write` to `mmu.h`, `mmu.c` & remove static mark,
> move `__gb_draw_line` to `cpu.h` ?

No, you don't have to move the above functions out of file `gameboy.h`. Instead, use [forward declaration](https://en.wikipedia.org/wiki/Forward_declaration) technique.

**Comment by Rsysz** [comment]
```c++
static uint8_t __gb_read(struct gb_s, const uint_fast16_t);
/* Internal function used to write bytes */
static void __gb_write(struct gb_s,
                       const uint_fast16_t,
                       const uint8_t);
#ifdef ENABLE_LCD
static void __gb_draw_line(struct gb_s);
#endif
```
Since they're static, `forward declaration` still take effect?

**Comment by jserv** [comment]
> Since they're static, `forward declaration` still take effect?

Simply remove `static` from the function implementation.

**Comment by Rsysz** [comment]
If i put `gb_run_frame` implementation at `cpu.c`,  `__gb_step_cpu` will not be linked, since it's static.
So, i just simply put it in `cpu.h`
```c++
void gb_run_frame(struct gb_s *gb)
{
    gb->gb_frame = 0;

    while (!gb->gb_frame)
        __gb_step_cpu(gb);
}
```

**Comment by jserv** [comment]
> If i put `gb_run_frame` implementation at `cpu.c`, `__gb_step_cpu` will not be linked, since it's static.
> So, i just simply put it in `cpu.h`

No, the problem resulted from insane `cpu.h`. The header file should provide function prototype and definitions. The real implementations should appear in `.c` files. You have to rework it first.

**Comment by Rsysz** [comment]
> > If i put `gb_run_frame` implementation at `cpu.c`, `__gb_step_cpu` will not be linked, since it's static.
> > So, i just simply put it in `cpu.h`
> 
> No, the problem resulted from insane `cpu.h`. The header file should provide function prototype and definitions. The real implementations should appear in `.c` files. You have to rework it first.

I move `__gb_step_cpu` , `__gb_execute_cb` and `gb_run_frame` implementation to `cpu.c`.
However, `bench.c` can't make through, so i wonder where `__gb_draw_line` symbol is located ?

File: cpu.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+#include <stddef.h>
+#include "core.h"
+
+uint8_t __gb_read(struct gb_s *gb, const uint_fast16_t addr);
+/* Internal function used to write bytes */
+void __gb_write(struct gb_s *gb,
+                const uint_fast16_t addr,
+                const uint8_t val);
+#ifdef ENABLE_LCD
```
Comment (jserv) [inline cpu.h]: Don't expose function `__gb_draw_line`.

Comment (jserv) [reply cpu.h]: Here, "expose" implies making something public.

File: cpu.c [hunk 4]
```diff
@@ -0,0 +1,2103 @@
...
+
+        if (gb->gb_reg.STAT & STAT_MODE_2_INTR)
+            gb->gb_reg.IF |= LCDC_INTR;
+    }
+    /* Update LCD */
+    else if (gb->lcd_mode == LCD_SEARCH_OAM &&
+             gb->counter.lcd_count >= LCD_MODE_3_CYCLES) {
+        gb->lcd_mode = LCD_TRANSFER;
+#if ENABLE_LCD
+        __gb_draw_line(gb);
```
Comment (jserv) [inline cpu.c:2101]: TODO: add forward declaration for `__gb_draw_line`.

**Comment by jserv** [comment]
> I move `__gb_step_cpu` , `__gb_execute_cb` and `gb_run_frame` implementation to `cpu.c`.
> However, `bench.c` can't make through, so i wonder where `__gb_draw_line` symbol is located ?

```diff
--- a/Makefile
+++ b/Makefile
@@ -52,7 +52,7 @@ $(OUT)/cpu_instrs.h: tests/cpu_instrs.gb tests/rom2h.c
 
 $(OUT)/bench: $(OUT)/cpu_instrs.h prof.h bench.c gameboy.h
        $(VECHO) "  CC+LD\t$@\n"
-       $(Q)$(CC) -o $@ bench.c
+       $(Q)$(CC) -DENABLE_LCD=0 -o $@ bench.c cpu.c
 
```

File: cpu.h [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <stddef.h>
+#include "core.h"
+
+uint8_t __gb_read(struct gb_s *gb, const uint_fast16_t addr);
```
Comment (jserv) [inline cpu.h]: Ditto. Don't expose `__gb_read` and `__gb_write`.

File: cpu.c [hunk 6]
```diff
@@ -0,0 +1,2103 @@
+#include "cpu.h"
+
```
Comment (jserv) [inline cpu.c:2]: TODO: add forward declaration for `__gb_read` and `__gb_write`.

**Comment by Rsysz** [comment]
Done ?

**Review by jserv — CHANGES_REQUESTED** [review]
Make bench program buildable and runnable.

**Comment by Rsysz** [comment]
> Make bench program buildable and runnable.

Done. Anything more?

File: Makefile [hunk 7]
```diff
@@ -51,7 +52,7 @@ $(OUT)/cpu_instrs.h: tests/cpu_instrs.gb tests/rom2h.c
 
 $(OUT)/bench: $(OUT)/cpu_instrs.h prof.h bench.c gameboy.h
 	$(VECHO) "  CC+LD\t$@\n"
-	$(Q)$(CC) -o $@ bench.c
+	$(Q)$(CC) -DENABLE_LCD=0 -o $@ bench.c cpu.c
```
Comment (jserv) [inline Makefile]: Change the sequence from `bench.c cpu.c` to `cpu.c bench.c`, which aids the compiler to understand the expected dependency.

File: cpu.h [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#pragma once
+
+#include <stddef.h>
```
Comment (jserv) [inline cpu.h]: Remove the inclusion of `stddef.h`.

Comment (Rsysz) [reply cpu.h]: Where should I put this header ?
Top of `cpu.c` ?

Comment (jserv) [reply cpu.h]: Move to header `core.h` if it is required.
You should always eliminate the inclusion of existing headers as possible.

**Comment by jserv** [comment]
Thank @Rsysz for contributing!
Let's move forward.
